scripts/wsrep_sst_xtrabackup-v2.sh

```diff
@@ -133,6 +136,90 @@ timeit(){
     return $extcode
 }
 
+# Helper function executed in background by interruptable_timeout function.
+# As it is executed in background it will return PIPESTATUS in the file
+# specified in argument.
+timeout_delegate(){
+    local stage=$1
+    shift
+    local status_file=$1
+    shift
+    local cmd="$@"
+    local x1 x2 took
+
+    if [[ $ttime -eq 1 ]];then 
+        x1=$(date +%s)
+        wsrep_log_info "Evaluating $cmd"
+        eval "$cmd; RC=( "\${PIPESTATUS[@]}" )"
+        x2=$(date +%s)
+        took=$(( x2-x1 ))
+        wsrep_log_info "NOTE: $stage took $took seconds"
+        totime=$(( totime+took ))
+    else 
+        wsrep_log_info "Evaluating $cmd"
+        eval "$cmd; RC=( "\${PIPESTATUS[@]}" )"
+    fi
+
+    echo ${RC[@]} > $status_file
+}
+
+# Execute provided command within provided time window.
+# Return:
+# Array consisting of error code of timeout followed by 
+# command PIPESTATUS array.
+# 0 - success
+# 124 - timeout
+interruptable_timeout() {
+    local tmt=$1
+    shift
+    local stage=$1
+    shift
+    local cmd="$@"
+    local result=0
+    local pid=
+
+    # Add 20 seconds for requested timeout to allow graceful shutdown.
+    tmt=$((tmt+20))
+
+    # Start the command in background and monitor if it is alive.
+    # If still alive after timeout, kill it.
+    timeout_delegate "$stage" "${dir}/$PIPESTATUS_FILE" "$cmd" &
+    pid=$!
+
+    while kill -0 $pid 2> /dev/null; do
+        if [[ $tmt -eq 20 ]];then
+          # Send SIGTERM to allow graceful shutdown
+          wsrep_log_info "Trying to terminate ($pid) $cmd with SIGTERM"
+          pkill -SIGTERM -P $pid
+          result=124
+        fi 
+
+        if [[ $tmt -eq 10 ]];then
+          # Still alive after SIGTERM? Send SIGKILL.
+          wsrep_log_info "Trying to terminate ($pid) $cmd with SIGKILL"
+          pkill -SIGKILL -P $pid
+          result=124
+        fi 
+        
+        if [[ $tmt -eq 0 ]];then
+          # Still alive after SIGKILL? Give up.
+          wsrep_log_error "Unable to stop command: $cmd"
+          result=124
+          break
+        fi 
+
+        sleep 1
+        tmt=$((tmt-1))
+    done
+    
+    if [[ $result -eq 0 ]];then
+      piperesult=(`cat ${dir}/$PIPESTATUS_FILE`)
+    fi
+
+    piperesult=( $result "${piperesult[@]}" )
```

Comment:
Doesn't piperesult need to be defined before this ('local piperesult')?

Reply:
Good point. The way it is now, piperesult becomes global variable, which is not necessary. 
I declared it local. 
I adjusted timeout_delegate() local variables in the same way.